storage/src/main/java/org/apache/kafka/server/log/internals/ProducerAppendInfo.java

```diff
@@ -0,0 +1,238 @@
+        return nextSeq == lastSeq + 1L || (nextSeq == 0 && lastSeq == Integer.MAX_VALUE);
+    }
+
+    public Optional<CompletedTxn> append(RecordBatch batch, Optional<LogOffsetMetadata> firstOffsetMetadataOpt) {
+        if (batch.isControlBatch()) {
+            Iterator<Record> recordIterator = batch.iterator();
+            if (recordIterator.hasNext()) {
+                Record record = recordIterator.next();
+                EndTransactionMarker endTxnMarker = EndTransactionMarker.deserialize(record);
+                return appendEndTxnMarker(endTxnMarker, batch.producerEpoch(), batch.baseOffset(), record.timestamp());
+            } else {
+                // An empty control batch means the entire transaction has been cleaned from the log, so no need to append
+                return Optional.empty();
+            }
+        } else {
+            LogOffsetMetadata firstOffsetMetadata = firstOffsetMetadataOpt.orElse(new LogOffsetMetadata(batch.baseOffset()));
+            appendDataBatch(batch.producerEpoch(), batch.baseSequence(), batch.lastSequence(), batch.maxTimestamp(),
+                    firstOffsetMetadata, batch.lastOffset(), batch.isTransactional());
+            return Optional.empty();
+        }
+    }
+
+    public void appendDataBatch(short epoch,
+                                int firstSeq,
+                                int lastSeq,
+                                long lastTimestamp,
+                                LogOffsetMetadata firstOffsetMetadata,
+                                long lastOffset,
+                                boolean isTransactional) {
+        long firstOffset = firstOffsetMetadata.messageOffset;
+        maybeValidateDataBatch(epoch, firstSeq, firstOffset);
+        updatedEntry.addBatch(epoch, lastSeq, lastOffset, (int) (lastOffset - firstOffset), lastTimestamp);
+
+        OptionalLong currentTxnFirstOffset = updatedEntry.currentTxnFirstOffset;
+        if (currentTxnFirstOffset.isPresent()) {
+            if (!isTransactional)
+                // Received a non-transactional message while a transaction is active
+                throw new InvalidTxnStateException("Expected transactional write from producer " + producerId + " at " +
+                        "offset " + firstOffsetMetadata + " in partition " + topicPartition);
+        } else {
+            if (isTransactional) {
+                // Began a new transaction
+                updatedEntry.currentTxnFirstOffset = OptionalLong.of(firstOffset);
+                transactions.add(new TxnMetadata(producerId, firstOffsetMetadata));
+            }
+        }
+    }
+
+    private void checkCoordinatorEpoch(EndTransactionMarker endTxnMarker, long offset) {
+        if (updatedEntry.coordinatorEpoch > endTxnMarker.coordinatorEpoch()) {
+            if (origin == AppendOrigin.REPLICATION) {
+                log.info("Detected invalid coordinator epoch for producerId " + producerId + " at " +
```

Comment:
We should use log placeholders here so that we avoid the string concat if the log is not enabled.